cmd/common/st/transfer.go

```diff
@@ -0,0 +1,307 @@
+	Run: func(cmd *cobra.Command, args []string) {
+		err := validateFlags()
+		if err != nil {
+			tracelog.ErrorLogger.FatalError(fmt.Errorf("invalid flags: %w", err))
+		}
+
+		sourceFolder, err := multistorage.ConfigureStorageFolder(transferSourceStorage)
+		if err != nil {
+			tracelog.ErrorLogger.FatalError(fmt.Errorf("can't configure source storage folder: %w", err))
+		}
+		targetFolder, err := multistorage.ConfigureStorageFolder(targetStorage)
+		if err != nil {
+			tracelog.ErrorLogger.FatalError(fmt.Errorf("can't configure target storage folder: %w", err))
+		}
+
+		parentDirPath := args[0]
+		err = transferFiles(sourceFolder, targetFolder, parentDirPath)
+		if err != nil {
+			tracelog.ErrorLogger.FatalError(err)
+		}
+	},
+}
+
+var (
+	transferSourceStorage            string
+	transferOverwrite                bool
+	transferConcurrency              int
+	transferMax                      int
+	transferAppearanceChecks         uint
+	transferAppearanceChecksInterval time.Duration
+)
+
+func init() {
+	transferCmd.Flags().StringVarP(&transferSourceStorage, "source", "s", "",
+		"storage name to move files from. Use 'default' to select the primary storage")
+	transferCmd.Flags().BoolVarP(&transferOverwrite, "overwrite", "o", false,
+		"whether to overwrite already existing files in the target storage and remove them from the source one")
+	transferCmd.Flags().IntVarP(&transferConcurrency, "concurrency", "c", 10,
+		"number of concurrent workers to move files. Value 1 turns concurrency off")
+	transferCmd.Flags().IntVarP(&transferMax, "max", "m", -1,
+		"max number of files to move in this run. Negative numbers turn the limit off")
+	transferCmd.Flags().UintVar(&transferAppearanceChecks, "appearance-checks", 3,
+		"number of times to check if a file is appeared for reading in the target storage after writing it. Value 0 turns checking off")
+	transferCmd.Flags().DurationVar(&transferAppearanceChecksInterval, "appearance-checks-interval", time.Second,
+		"minimum time interval between performing checks for files to appear in the target storage")
+
+	StorageToolsCmd.AddCommand(transferCmd)
+}
+
+func validateFlags() error {
+	if transferSourceStorage == "" {
+		return fmt.Errorf("source storage must be specified")
+	}
+	if transferSourceStorage == "all" {
+		return fmt.Errorf("an explicit source storage must be specified instead of 'all'")
+	}
+	if targetStorage == "all" {
+		return fmt.Errorf("an explicit target storage must be specified instead of 'all'")
+	}
+	if transferSourceStorage == targetStorage {
+		return fmt.Errorf("source and target storages must be different")
+	}
+	if transferConcurrency < 1 {
+		return fmt.Errorf("concurrency level must be >= 1 (which turns it off)")
+	}
+	return nil
+}
+
+func transferFiles(source, target storage.Folder, parentDirPath string) error {
+	files, err := listFilesToMove(source, target, parentDirPath)
+	if err != nil {
+		return err
+	}
+
+	fileStates := make(chan transferFileState, len(files))
```

Comment:
OK, I'll try to make it so. Here's just a small remark: I'm not sure we have a lot of stuff to do in the constructor. For example, we can't really init workers because we don't know their number before listing the files. It's not really necessary to init more workers than we have files to move. 

And I'll introduce clearer steps of files processing, OK